lib/reconcile/astarte_priorityclass.go

```diff
@@ -0,0 +1,130 @@
+/*
+  This file is part of Astarte.
+
+  Copyright 2021 Ispirata Srl
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package reconcile
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	scheduling "k8s.io/api/scheduling/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	controllerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	apiv1alpha1 "github.com/astarte-platform/astarte-kubernetes-operator/apis/api/v1alpha1"
+	"github.com/astarte-platform/astarte-kubernetes-operator/lib/misc"
+)
+
+const (
+	AstarteHighPriorityName string = "astarte-high-priority-non-preemptive"
+	AstarteMidPriorityName  string = "astarte-mid-priority-non-preemptive"
+	AstarteLowPriorityName  string = "astarte-low-priority-non-preemptive"
+)
+
+var defaultComponentPriorityClass = map[apiv1alpha1.AstarteComponent]string{
+	apiv1alpha1.AppEngineAPI:       AstarteMidPriorityName,
+	apiv1alpha1.DataUpdaterPlant:   AstarteHighPriorityName,
+	apiv1alpha1.FlowComponent:      AstarteMidPriorityName,
+	apiv1alpha1.Housekeeping:       AstarteLowPriorityName,
+	apiv1alpha1.HousekeepingAPI:    AstarteLowPriorityName,
+	apiv1alpha1.Pairing:            AstarteMidPriorityName,
+	apiv1alpha1.PairingAPI:         AstarteMidPriorityName,
+	apiv1alpha1.RealmManagement:    AstarteLowPriorityName,
+	apiv1alpha1.RealmManagementAPI: AstarteLowPriorityName,
+	apiv1alpha1.TriggerEngine:      AstarteLowPriorityName,
+	apiv1alpha1.Dashboard:          AstarteLowPriorityName,
+}
+
+func EnsureAstartePriorityClasses(instance *apiv1alpha1.Astarte, c client.Client, scheme *runtime.Scheme) error {
+	//Shall we use priorityClasses?
+	if instance.Spec.Features.AstartePodPriorities.IsEnabled() {
+
+		//we don't want to preempt other pods
+		preemptionPolicy := v1.PreemptNever
+
+		// default value makes sure this pointer is not nil
+		highPriorityValue := int32(*instance.Spec.Features.AstartePodPriorities.AstarteHighPriority)
+
+		//High priorityClass
+		priorityClassHigh := &scheduling.PriorityClass{
+			ObjectMeta:       metav1.ObjectMeta{Name: AstarteHighPriorityName},
+			Value:            highPriorityValue,
+			GlobalDefault:    false,
+			Description:      "Astarte high-priority pods (e.g. RabbitMQ, VerneMQ, Astarte Data Updater Plant) should be in this priority class.",
+			PreemptionPolicy: &preemptionPolicy,
```

Comment:
I'd expect that these fields are set within CreateOrUpdate mutate function. (Same for the mid and low classes)